script.lua

```diff
@@ -6349,37 +6337,87 @@ function setupColor(table, color)
         colorTint = Color.fromHex(Tints[color].Presence)
     end
     table.setColorTint(colorTint)
-    table.UI.setXml("")
     table.clearButtons()
-    setupSwapButtons(table)
+    setupSwapButtons(table, { seated = true })
 
     updateColorPickButtons()
-    updateSwapButtons()
 end
-function setupSwapButtons(obj)
-    local scale = flipVector(Vector(obj.getScale()))
-    scale = scale * 2
-    -- Swap Place (button index 0)
-    obj.createButton({
-        label="", click_function="onClickedSwapPlace", function_owner=Global,
-        position={-3.25,0.4,7.5}, rotation={0,0,0}, height=0, width=0, scale=scale,
-        font_color={0,0,0}, font_size=250,
-        tooltip="Moves your current player color to be located here. The color currently seated here will be moved to your current location. Spirit panels and other cards will be relocated if applicable.",
+-- @param params.seated Boolean indicating if a player is about to be seated here.
+function setupSwapButtons(obj, params)
+    params = params or {}
+    local xml = {}
+    local color = getTableColor(obj)
+    local buttonColor = Color[color]:toHex(false)
+    local textColor = fontColor(Color[color])
+    local playerButtonVisibility = "Invisible"
+    if showPlayerButtons then
+        playerButtonVisibility = visiTableToString(invertVisiTable({"Black", "Grey", color}))
+    end
+    local playSpiritVisibility = "Invisible"
+    local playSpiritText = ""
+    if not params.seated and not Player[color].seated then
+        if selectedColors[color] then
+            playSpiritVisibility = visiTableToString(invertVisiTable({color}))
+            playSpiritText = "Play Spirit"
+        elseif showPlayerButtons then
+            playSpiritVisibility = visiTableToString(invertVisiTable({color}))
+            playSpiritText = "Play " .. color
+        end
+    end
+    table.insert(xml, {
+        tag = "Button",
+        attributes = {
+            id = "swapPlace",
+            onClick = "Global/onClickedSwapPlace("..obj.guid..")",
+            position = "320 755 -80",
+            rotation = "0 0 180",
+            text = "Swap Place",
+            fontSize = "44",
+            width = "270",
+            height = "110",
+            visibility = playerButtonVisibility,
+            -- Note: tooltips don't work on Custom UI (https://tabletopsimulator.nolt.io/1369)
```

Comment:
tooltips do work for non global scripts, setup checker xml for example uses them

Reply:
Hmm. I'm not sure exactly what is going on. But it appears that the behaviour of custom UI may depend on the type of object (or other factors). It isn't that tooltips don't work at all, but rather that at least in this instance, they are rendered in a way that is obscured by the object they are on.

If you enable the tooltips here, you can see part of the tooltip for the "Swap Places" button on the leftmost place sticking out the far left side of the table. I tried a bunch of things to try and figure out how to get the tooltip to render enough above the table to be visible, but did not have any luck.

After some more investigation, it looks like the tooltip is always drawn right on the surface of the object the UI is attached to, regardless of the height of the UI element. In the case of the setup checker, this is an object floating above the table, so the tooltip doesn't intersect with the table. That doesn't work for UI that is attached to a player table.

Reply:
if the issue is the tooltip getting drawn on the right side of the object you can just change that with the tooltipPosition attribute

Reply:
The issue is the tooltip getting drawn *inside* the table, so it isn't visible.

Reply:
As far as I can tell, `tooltipPosition` only affects the horizontal position of the tooltip, but the issue is the vertical height of the tooltip.

Reply:
the position -80 seems a bit sus, if you really need that, try wrapping the buttons inside of a panel that contains the -80 offset

Reply:
I copied that value from `setupColorPickButtons`.

Reply:
did wrapping the buttons inside a panel not fix the tooltip issue?

Reply:
No. You can see this clearly, if you set one of the panels on `SetupChecker` to be at z position -1000 (for example), and notice that the tooltips for those objects are still right near the table (rather than suspended way in the air).

Reply:
did a little more testing on this, and seems tooltip's vertical height is based off of the object it's attached to and ignores everything else